pypose/module/ICP.py

```diff
@@ -0,0 +1,140 @@
+            while iter_num <= self.max_iterations:
+                iter_num += 1
+                distances, indices = self.nearest_neighbor(temp_pc, p2)
+                mean_err = sum(distances) / len(distances)
+                transR, transT = self.get_transform(temp_pc, p2[indices])
+                temp_pc = torch.matmul(temp_pc, transpose(transR)) + transT
+                if (abs(prev_err - mean_err) < self.tolerance):
+                    break
+                prev_err = mean_err
+            transR, transT = self.get_transform(p1, temp_pc)
+            return self.RT2SE3(transR, transT)
+        else:
+            transR, transT = self.get_transform(p1, p2)
+            return self.RT2SE3(transR, transT)
+
+    def RT2SE3(self, transR: torch.tensor, transT: torch.tensor):
+        r'''
+        Transfers the transR and transT into SE3 Lie tensor
+
+        Args:
+            transR: the rotation matrix 3x3
+            transT: the translation matrix 3x1
+
+        Returns:
+            transformation matrix LieTensor SE3
+        '''
+
+        temp_transT = transT.reshape([3, 1])
+        T = torch.cat([torch.cat([transR, temp_transT], dim=1), torch.tensor([[0., 0., 0., 1.]], dtype=torch.float)])
+        return pp.from_matrix(T, ltype=pp.SE3_type)
+
+    def get_transform(self, p1: torch.tensor, p2: torch.tensor):
+        r'''Using SVD algorithm to calculate the transformation matrix between the corresponding points set p1, p2.
+
+        Args:
+            p1: input points set
+            p2: the matched target points set
+
+        Returns:
+            R: the rotation matrix between p1 and p2
+            T: the translation matrix between p1 and p2
+        '''
+        p1_centroid = p1.mean(0)
+        p2_centroid = p2.mean(0)
+
+        temp_p1 = (p1 - p1_centroid)
+        temp_p2 = (p2 - p2_centroid)
+
+        H = torch.matmul(transpose(temp_p2), temp_p1)
+        u, s, vT = torch.linalg.svd(H)
+        v = transpose(vT)
+        uT = transpose(u)
+
+        if (torch.det(v * uT) < 0):
+            v[2, :] *= -1
+        R = torch.matmul(u, vT)
+        t = p2_centroid - torch.matmul(p1_centroid, transpose(R))
+
+        return R, t
+
+
+    def nearest_neighbor(self, p1, p2):
+        r'''
+        Select the nearest neighbor point of p1 from p2
+        Args:
+            p1: the source points set
+            p2: the target points set
+            tolerance: the threshold of min distance
+
+        Returns:
+            distances: the min distance between point in p1 and its nearest neighbor
+            indices: the index of the nearest neighbor point in p2
+        '''
+
+        distances = torch.zeros([p1.size()[0], 1])
+        indices = list(range(p1.size()[0]))
+        for i in range(p1.size()[0]):
+            dist1 = torch.norm(p2-p1[i],dim=1,p=None)
+            nn = dist1.topk(1, largest = False)
+            distances[i] = nn.values[0]
+            indices[i] = int(nn.indices[0])
```

Comment:
don't use loop to find nearest neighbor, something like `(p2.unsqueeze(-1) - p1).norm(-1).argmax()` should work.